Sources/StreamChat/Workers/ChatClientUpdater.swift

```diff
@@ -11,110 +11,78 @@ class ChatClientUpdater {
         self.client = client
     }
 
+    private enum EnvironmentState {
+        case firstConnection
+        case newToken
+        case newUser
+    }
+
     func prepareEnvironment(
         userInfo: UserInfo?,
         newToken: Token,
         completion: @escaping (Error?) -> Void
     ) {
-        guard let currentUserId = client.currentUserId else {
-            // Set the current user id
-            client.currentUserId = newToken.userId
-            // Set the token
-            client.currentToken = newToken
-            // Set the web-socket endpoint
-            client.webSocketClient?.connectEndpoint = .webSocketConnect(userInfo: userInfo ?? .init(id: newToken.userId))
-            // Create background workers
-            client.createBackgroundWorkers()
-            // Provide the token to pending API requests
-            client.completeTokenWaiters(token: newToken)
-            completion(nil)
-            return
+        let state: EnvironmentState
+        if client.currentUserId == nil {
+            state = .firstConnection
+        } else if newToken.userId == client.currentUserId {
+            state = .newToken
+        } else {
+            state = .newUser
         }
-        
-        guard newToken.userId == currentUserId else {
-            // Cancel all API requests since they are related to the previous user.
-            client.completeTokenWaiters(token: nil)
 
-            // Setting a new user is not possible in connectionless mode.
-            guard client.config.isClientInActiveMode else {
-                completion(ClientError.ClientIsNotInActiveMode())
+        switch state {
+        case .firstConnection:
+            client.updateWebSocketEndpoint(with: newToken, userInfo: userInfo)
+            client.updateUser(with: newToken, completeTokenWaiters: true, isFirstConnection: true)
+            completion(nil)
+
+        case .newToken:
+            client.updateWebSocketEndpoint(with: newToken, userInfo: userInfo)
+
+            guard newToken != client.currentToken else {
+                completion(nil)
                 return
             }
 
-            // Update the current user id to the new one.
-            client.currentUserId = newToken.userId
+            client.updateUser(with: newToken, completeTokenWaiters: true, isFirstConnection: false)
+            completion(nil)
+
+        case .newUser:
+            client.switchToNewUser(with: newToken)
 
-            // Update the current token with the new one.
-            client.currentToken = newToken
+            // Setting a new connection is not possible in connectionless mode.
+            guard client.config.isClientInActiveMode else {
+                completion(ClientError.ClientIsNotInActiveMode())
+                return
+            }
```

Comment:
This is nice! This means it would also fix this ticket right: https://stream-io.atlassian.net/browse/CIS-2208

Probably worth adding it to the PR title, and PR Description in issue links

Reply:
I would not say this guarentees a fix for this issue, but will review it in subsequent PRs when I also tackle https://stream-io.atlassian.net/browse/CIS-1727